tools/gap2seq.py

```diff
@@ -0,0 +1,110 @@
+
+        Params:
+            solid_kmer_thresholds: kmers must appear at least this many times in the reads to be considered solid.
+                We try gapfilling for all combinations of values of solid_kmer_thresholds and kmer_sizes.
+            kmer_sizes: kmer sizes to use.  We try gapfilling for all combinations of values of solid_kmer_thresholds and kmer_sizes.
+            min_gap_to_close: stop gap-closing if all gaps are no longer than this many Ns
+            gap2seq_opts: extra command-line flags to pass to Gap2Seq
+            mem_limit_gb: max memory to use, in gigabytes
+            threads: number of threads to use
+            time_limit_minutes: stop trying to close more gaps after this many minutes (currently this is a soft/advisory limit)
+        
+        """
+        from os.path import abspath, join
+        solid_kmer_thresholds = sorted(util.misc.make_seq(solid_kmer_thresholds), reverse=True)
+        kmer_sizes = sorted(util.misc.make_seq(kmer_sizes), reverse=True)
+        samtools = tools.samtools.SamtoolsTool()
+        with samtools.bam2fq_tmp(inBam) as (reads1, reads2):
+            done = False
+            prev_scaffold = in_scaffold
+            stop_time = time.time() + 60*time_limit_minutes
+            with util.file.tmp_dir('_gap2seq_dir') as gap2seq_dir:
+                for solid_kmer_threshold in solid_kmer_thresholds:
+                    for kmer_size in kmer_sizes:
+                        gap2seq_filled = join(gap2seq_dir, 'gap2seq-filled.s{}.k{}.fasta'.format(solid_kmer_threshold, kmer_size))
+                        log.debug('s={} k={} gap2seq_filled={}'.format(solid_kmer_threshold, kmer_size, gap2seq_filled))
+
+                        # gap2seq (or rather the gatb library it uses) currently has a bug where tempfiles are left in the
+                        # current working directory.  So we run it in its own temp dir, but then must give absolute paths
+                        # for all files.
+
+                        args = ['-scaffolds', abspath(prev_scaffold), '-filled', abspath(gap2seq_filled),
+                                '-reads', ','.join((abspath(reads1), abspath(reads2))), '-solid', str(solid_kmer_threshold), 
+                                '-k', str(kmer_size), '-all-upper',
+                                '-verbose', '-nb-cores', str(threads), '-max-mem', str(mem_limit_gb)] + shlex.split(gap2seq_opts)
+                        
+                        with util.file.tmp_dir('_gap2seq_run_dir') as gap2seq_run_dir:
+                            with util.file.tmp_chdir(gap2seq_run_dir):
+                                self.execute(args=args)
+
+                        prev_scaffold = gap2seq_filled
+
+                        if not any('N'*min_gap_to_close in str(rec.seq) for rec in Bio.SeqIO.parse(gap2seq_filled, 'fasta')):
+                            log.info('no gaps left, quittting gap2seq early')
```

Comment:
Right.  Sometimes we can tell that won't help, e.g. if the gaps that are left are too large.  There are other pipeline places where we could also suggest what to try tuning.